File: GateSyntaxTS/src/runtime/syntax_parser.ts
// Line-oriented parser — mirrors GateSyntax.Runtime.SyntaxParser.cs
import {
  SyntaxNode, ElementDecl, StateDecl, Property, Behavior,
  ValueExpr, LiteralExpr, RefExpr, BinaryExpr,
  literal, ref, binary,
} from './syntax_node';

const SEP = ' :: ';

export class SyntaxParser {
  parseContent(content: string, sourceName = 'inline.ui'): SyntaxNode[] {
    void sourceName;
    return content
      .split('\n')
      .map(l => l.trim())
      .filter(l => l && !l.startsWith('//') && !l.startsWith('#'))
      .flatMap(l => {
        const node = this.parseLine(l);
        return node ? [node] : [];
      });
  }

  private parseLine(line: string): SyntaxNode | null {
    const tokens = line.split(SEP);
    if (!tokens.length) return null;
    return tokens[0].trimStart().startsWith('/')
      ? this.parseStateDecl(tokens)
      : this.parseElementDecl(tokens);
  }

  private parseStateDecl(tokens: string[]): StateDecl {
    const name = tokens[0].trimStart().replace(/^\//, '').trim();
    const raw   = tokens[1]?.replace(/\\$/, '').trim() ?? '';
    const saved = tokens.slice(2).some(t => t.trim().toUpperCase() === 'SAVED');
    return { kind: 'state', name, defaultValue: SyntaxParser.parseLiteral(raw), saved };
  }

  private parseElementDecl(tokens: string[]): ElementDecl {
    const first = tokens[0].trim();
    const sp    = first.indexOf(' ');
    const noun  = (sp < 0 ? first : first.slice(0, sp)).toUpperCase();
    const id    = sp < 0 ? first : first.slice(sp + 1).trim();

    const props: Property[] = [];
    const behaviors: Behavior[] = [];

    let i = 1;
    while (i < tokens.length) {
      const seg = tokens[i].trim();
      if (seg.toUpperCase().startsWith('ON ')) {
        const parts = seg.split(/\s+/);
        if (parts.length >= 3) {
          const eventName = parts[1].toUpperCase();
          const targetVar = parts[2].replace(/^\//, '').trim();
          const expr = tokens[i + 1]?.replace(/\\$/, '').trim() ?? '';
          i++;
          behaviors.push({ event: eventName, targetVar, expression: expr });
        } else if (parts.length === 2) {
          const eventName = parts[1].toUpperCase();
          const expr = tokens[i + 1]?.replace(/\\$/, '').trim() ?? '';
          i++;
          behaviors.push({ event: eventName, targetVar: '__noop__', expression: expr });
        }
      } else {
        const sp2 = seg.indexOf(' ');
        if (sp2 < 0) {
          props.push({ key: seg.toUpperCase(), value: literal(true) });
        } else {
          const key = seg.slice(0, sp2).toUpperCase();
          const val = seg.slice(sp2 + 1).trim();
          props.push({ key, value: SyntaxParser.parseValueExpr(val) });
        }
      }
      i++;
    }

    return { kind: 'element', noun, id, props, behaviors };
  }

  static parseValueExpr(s: string): ValueExpr {
    const parts = SyntaxParser.tokenizeExpr(s.trim());
    if (!parts.length) return literal('');
    if (parts.length === 1) return SyntaxParser.parseSingleToken(parts[0]);
    let left = SyntaxParser.parseSingleToken(parts[0]);
    for (let i = 1; i < parts.length - 1; i += 2) {
      left = binary(left, parts[i], SyntaxParser.parseSingleToken(parts[i + 1]));
    }
    return left;
  }

  private static tokenizeExpr(s: string): string[] {
    const result: string[] = [];
    let buf = '';
    let inQuote = false;
    let inRef = false;
    for (const ch of s) {
      if      (ch === '"')              { inQuote = !inQuote; buf += ch; }
      else if (ch === '[')              { inRef = true;  buf += ch; }
      else if (ch === ']')              { inRef = false; buf += ch; }
      else if (ch === ' ' && !inQuote && !inRef) { if (buf) { result.push(buf); buf = ''; } }
      else                              { buf += ch; }
    }
    if (buf) result.push(buf);
    return result;
  }

  private static parseSingleToken(t: string): LiteralExpr | RefExpr {
    t = t.trim();
    if (t.startsWith('[') && t.endsWith(']')) return ref(t.slice(1, -1));
    return literal(SyntaxParser.parseLiteral(t));
  }

  static parseLiteral(s: string): unknown {
    s = s.trim().replace(/\\$/, '').trim();
    if (s.startsWith('"') && s.endsWith('"')) return s.slice(1, -1);
    if (s.toUpperCase() === 'TRUE')  return true;
    if (s.toUpperCase() === 'FALSE') return false;
    const n = Number(s);
    if (!isNaN(n) && s !== '') return n;
    return s;
  }
}

// Re-export for external consumers
export type { LiteralExpr, RefExpr, BinaryExpr };

File: GateSyntaxJava/core/com/gatesyntax/runtime/SyntaxParser.java
package com.gatesyntax.runtime;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Line-oriented parser — mirrors GateSyntax.Runtime.SyntaxParser. */
public class SyntaxParser {

    private static final String SEP = " :: ";

    public List<SyntaxNode> parseContent(String content, String sourceName) {
        return content.lines()
                .map(String::strip)
                .filter(l -> !l.isEmpty() && !l.startsWith("//") && !l.startsWith("#"))
                .flatMap(l -> parseLineOpt(l).stream())
                .toList();
    }

    // ── Internal ──────────────────────────────────────────────────────────────

    private java.util.Optional<SyntaxNode> parseLineOpt(String line) {
        String[] tokens = line.split(java.util.regex.Pattern.quote(SEP), -1);
        if (tokens.length == 0) return java.util.Optional.empty();
        SyntaxNode node = tokens[0].stripLeading().startsWith("/")
                ? parseStateDecl(tokens)
                : parseElementDecl(tokens);
        return java.util.Optional.ofNullable(node);
    }

    private StateDecl parseStateDecl(String[] tokens) {
        String name = tokens[0].stripLeading().replaceFirst("^/", "").strip();
        String raw  = tokens.length > 1 ? tokens[1].replaceAll("\\\\$", "").strip() : "";
        boolean saved = Arrays.stream(tokens).skip(2).anyMatch(t -> t.strip().equalsIgnoreCase("SAVED"));
        return new StateDecl(name, parseLiteral(raw), saved);
    }

    private ElementDecl parseElementDecl(String[] tokens) {
        String first = tokens[0].strip();
        int sp = first.indexOf(' ');
        String noun = (sp < 0 ? first : first.substring(0, sp)).toUpperCase();
        String id   = sp < 0 ? first : first.substring(sp + 1).strip();

        List<Property> props     = new ArrayList<>();
        List<Behavior> behaviors = new ArrayList<>();

        int i = 1;
        while (i < tokens.length) {
            String seg = tokens[i].strip();
            if (seg.toUpperCase().startsWith("ON ")) {
                String[] parts = seg.split("\\s+");
                if (parts.length >= 3) {
                    String event  = parts[1].toUpperCase();
                    String target = parts[2].replaceFirst("^/", "").strip();
                    String expr   = i + 1 < tokens.length ? tokens[i + 1].replaceAll("\\\\$", "").strip() : "";
                    i++;
                    behaviors.add(new Behavior(event, target, expr));
                } else if (parts.length == 2) {
                    String event = parts[1].toUpperCase();
                    String expr  = i + 1 < tokens.length ? tokens[i + 1].replaceAll("\\\\$", "").strip() : "";
                    i++;
                    behaviors.add(new Behavior(event, "__noop__", expr));
                }
            } else {
                int sp2 = seg.indexOf(' ');
                if (sp2 < 0) {
                    props.add(new Property(seg.toUpperCase(), new LiteralExpr(true)));
                } else {
                    String key = seg.substring(0, sp2).toUpperCase();
                    String val = seg.substring(sp2 + 1).strip();
                    props.add(new Property(key, parseValueExpr(val)));
                }
            }
            i++;
        }
        return new ElementDecl(noun, id, List.copyOf(props), List.copyOf(behaviors));
    }

    // ── Value expression parsing ──────────────────────────────────────────────

    public static ValueExpr parseValueExpr(String s) {
        List<String> parts = tokenizeExpr(s.strip());
        if (parts.isEmpty()) return new LiteralExpr("");
        if (parts.size() == 1) return parseSingleToken(parts.get(0));
        ValueExpr left = parseSingleToken(parts.get(0));
        for (int i = 1; i < parts.size() - 1; i += 2) {
            left = new BinaryExpr(left, parts.get(i), parseSingleToken(parts.get(i + 1)));
        }
        return left;
    }

    private static List<String> tokenizeExpr(String s) {
        List<String> result = new ArrayList<>();
        StringBuilder buf = new StringBuilder();
        boolean inQuote = false, inRef = false;
        for (char ch : s.toCharArray()) {
            if      (ch == '"')                    { inQuote = !inQuote; buf.append(ch); }
            else if (ch == '[')                    { inRef = true;  buf.append(ch); }
            else if (ch == ']')                    { inRef = false; buf.append(ch); }
            else if (ch == ' ' && !inQuote && !inRef) {
                if (!buf.isEmpty()) { result.add(buf.toString()); buf.setLength(0); }
            }
            else                                   { buf.append(ch); }
        }
        if (!buf.isEmpty()) result.add(buf.toString());
        return result;
    }

    private static ValueExpr parseSingleToken(String t) {
        t = t.strip();
        if (t.startsWith("[") && t.endsWith("]")) return new RefExpr(t.substring(1, t.length() - 1));
        return new LiteralExpr(parseLiteral(t));
    }

    static Object parseLiteral(String s) {
        s = s.strip().replaceAll("\\\\$", "").strip();
        if (s.startsWith("\"") && s.endsWith("\"")) return s.substring(1, s.length() - 1);
        if (s.equalsIgnoreCase("TRUE"))  return Boolean.TRUE;
        if (s.equalsIgnoreCase("FALSE")) return Boolean.FALSE;
        try { return Integer.parseInt(s); }   catch (NumberFormatException ignored) {}
        try { return Double.parseDouble(s); } catch (NumberFormatException ignored) {}
        return s;
    }
}
